File: Engine/MapEditor/mapelement.cpp
/*
    RPG Paper Maker Copyright (C) 2017 Marie Laporte

    This file is part of RPG Paper Maker.

    RPG Paper Maker is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    RPG Paper Maker is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "mapelement.h"

QString MapElement::jsonOrientation = "o";
QString MapElement::jsonUp = "up";
QString MapElement::jsonX = "xOff";
QString MapElement::jsonY = "yOff";
QString MapElement::jsonZ = "zOff";

// -------------------------------------------------------
//
//  CONSTRUCTOR / DESTRUCTOR / GET / SET
//
// -------------------------------------------------------

MapElement::MapElement() :
    m_orientation(OrientationKind::None),
    m_up(CameraUpDownKind::None),
    m_xOffset(0),
    m_yOffset(0),
    m_zOffset(0)
{

}

MapElement::~MapElement()
{

}

void MapElement::setUpDown(CameraUpDownKind upDown) {
    m_up = upDown;
}

MapEditorSelectionKind MapElement::getKind() const{
    return MapEditorSelectionKind::Land;
}

MapEditorSubSelectionKind MapElement::getSubKind() const{
    return MapEditorSubSelectionKind::None;
}

// -------------------------------------------------------
//
//  READ / WRITE
//
// -------------------------------------------------------

void MapElement::read(const QJsonObject & json){
    if (json.contains(MapElement::jsonOrientation)) {
        m_orientation = static_cast<OrientationKind>(
                    json[MapElement::jsonOrientation].toInt());
    }
    if (json.contains(MapElement::jsonUp))
        m_up = static_cast<CameraUpDownKind>(json[MapElement::jsonUp].toInt());
    if (json.contains(MapElement::jsonX))
        m_xOffset = json[MapElement::jsonX].toInt();
    if (json.contains(MapElement::jsonY))
        m_yOffset = json[MapElement::jsonY].toInt();
    if (json.contains(MapElement::jsonZ))
        m_zOffset = json[MapElement::jsonZ].toInt();
}

// -------------------------------------------------------

void MapElement::write(QJsonObject &json) const{
    if (m_orientation != OrientationKind::None)
        json[MapElement::jsonOrientation] = (int) m_orientation;
    if (m_up != CameraUpDownKind::None)
        json[MapElement::jsonUp] = (int) m_up;
    if (m_xOffset != 0)
        json[MapElement::jsonX] = m_xOffset;
    if (m_yOffset != 0)
        json[MapElement::jsonY] = m_yOffset;
    if (m_zOffset != 0)
        json[MapElement::jsonZ] = m_zOffset;
}

File: Game/Content/Datas/Scripts/System/map-element.js
/*
    RPG Paper Maker Copyright (C) 2017 Marie Laporte

    This file is part of RPG Paper Maker.

    RPG Paper Maker is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    RPG Paper Maker is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
*/

// -------------------------------------------------------
//
//  CLASS MapElement
//
// -------------------------------------------------------

/** @class
*   An element in the map.
*   @property {Orientation} orientation The orientation according to layer.
*   @property {CameraUpDown} upDown The camera up down orientation according to
*   layer.
*   @property {number} xOffset The offset of the object according to layer.
*   @property {number} yOffset The offset of the object according to layer.
*   @property {number} zOffset The offset of the object according to layer.
*/
function MapElement() {
    this.xOffset = 0;
    this.yOffset = 0;
    this.zOffset = 0;
}

MapElement.prototype = {

    /** Read the JSON associated to the map element.
    *   @param {Object} json Json object describing the object.
    */
    read: function(json) {
        var x = json.xOff;
        var y = json.yOff;
        var z = json.zOff;

        if (typeof(x) !== 'undefined')
            this.xOffset = x;
        if (typeof(y) !== 'undefined')
            this.yOffset = y;
        if (typeof(z) !== 'undefined')
            this.zOffset = z;
    },

    /** Scale the vertices correctly.
    *   @param {THREE.Vector3} vecA The A vertex to rotate.
    *   @param {THREE.Vector3} vecB The B vertex to rotate.
    *   @param {THREE.Vector3} vecC The C vertex to rotate.
    *   @param {THREE.Vector3} vecD The D vertex to rotate.
    *   @param {THREE.Vector3} center The center to rotate around.
    */
    scale: function(vecA, vecB, vecC, vecD, center, position, size, kind) {
        var zPlus = RPM.positionLayer(position) * 0.05;

        // Apply an offset according to layer position
        if (kind !== ElementMapKind.SpritesFace && !this.front)
            zPlus *= -1;

        var offset = new THREE.Vector3(0, 0, zPlus);

        // Center
        center.setX(this.xOffset * $SQUARE_SIZE);
        center.setY(this.yOffset * $SQUARE_SIZE);
        center.setZ(this.zOffset * $SQUARE_SIZE);

        // Position
        var pos = center.clone();
        pos.add(offset);
        vecA.multiply(size);
        vecB.multiply(size);
        vecC.multiply(size);
        vecD.multiply(size);
        vecA.add(pos);
        vecB.add(pos);
        vecC.add(pos);
        vecD.add(pos);
    }
}
